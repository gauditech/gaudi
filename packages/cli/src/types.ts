/**
 * Something that can be stopped like eg. resource watcher.
 *
 * We should add "start" if we need to have control over starting (eg. do it later/elsewhere).
 */
export type Stoppable = {
  stop: () => Promise<void>;
};
