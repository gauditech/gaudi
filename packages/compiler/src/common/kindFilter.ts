type Kind = string | number | symbol | boolean | undefined | null;

export type FilteredByKind<T extends { kind: Kind }, kind = T["kind"]> = kind extends T["kind"]
  ? Exclude<{ [K in keyof T]: K extends "kind" ? kind : T[K] }, { kind: never }>
  : never;

export type RejectedByKind<T extends { kind: Kind }, kind = T["kind"]> = Exclude<
  FilteredByKind<T>,
  { kind: kind }
>;

export function kindFilter<i extends { kind: Kind }[], const k extends i[number]["kind"]>(
  input: i,
  ...kinds: k[]
): FilteredByKind<i[number], k>[] {
  return input.filter((i) => {
    for (const kind of kinds) {
      if (kind === i.kind) return true;
    }
    return false;
  }) as FilteredByKind<i[number], k>[];
}

export function kindReject<i extends { kind: Kind }[], const k extends i[number]["kind"]>(
  input: i,
  ...kinds: k[]
): RejectedByKind<i[number], k>[] {
  return input.filter((i) => {
    for (const kind of kinds) {
      if (kind !== i.kind) return true;
    }
    return false;
  }) as RejectedByKind<i[number], k>[];
}

export function kindFind<i extends { kind: Kind }[], const k extends i[number]["kind"]>(
  input: i,
  ...kinds: k[]
): FilteredByKind<i[number], k> | undefined {
  return input.find((i) => {
    for (const kind of kinds) {
      if (kind === i.kind) return true;
    }
    return false;
  }) as FilteredByKind<i[number], k> | undefined;
}
