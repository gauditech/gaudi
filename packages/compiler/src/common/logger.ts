import { AsyncLocalStorage } from "node:async_hooks";
import util from "util";

import chalk from "chalk";
import _ from "lodash";
import { createLogger, format, transports } from "winston";

const als = new AsyncLocalStorage<string[]>();

/**
 * Logger implementation
 */

const winstonLogger = createLogger({
  level: "silly",
  format: format.combine(
    format.timestamp({
      format: "YYYY-MM-DD HH:mm:ss",
    }),
    format.label(),
    format.errors({ stack: true }),
    format.splat(),
    format.align(),
    format.padLevels(),
    format.printf(({ level, label, message, timestamp, stack }) => {
      const cStack = coloredStack(stack);
      const cTimestamp = coloredTs(timestamp);
      const cLevel = coloredLevel(level);
      const cLabel = coloredLabel(label);
      const msg = stack ? "" : padMessage(message);

      return `${cTimestamp} ${cLevel}${cLabel} ${msg}${cStack}`;
    })
  ),
  transports: [new transports.Console({})],
});

type LogLevel = "error" | "warn" | "info" | "verbose" | "debug" | "silly";

interface LogConfig {
  [category: string]: LogLevel;
}

const LOG_CATEGORIES: LogConfig = {
  "": "silly",
  sql: "debug",
  http: "debug",
  debug: "silly",
};

type Loggable = unknown;

export class Logger {
  private _category: string | null = null;
  private _forceLevel: LogLevel | undefined;

  constructor(category?: string, level?: LogLevel) {
    this._category = category ?? null;
    this._forceLevel = level;
  }

  silly(label: Loggable, message: Loggable): void;
  silly(message: Loggable): void;
  silly(...args: Loggable[]) {
    if (args.length > 1) {
      this._log("silly", args[1], args[0]);
    } else {
      // only message is passed
      this._log("silly", args[0], undefined);
    }
  }

  debug(label: Loggable, message: Loggable): void;
  debug(message: Loggable): void;
  debug(...args: Loggable[]) {
    if (args.length > 1) {
      this._log("debug", args[1], args[0]);
    } else {
      // only message is passed
      this._log("debug", args[0], undefined);
    }
  }

  info(label: Loggable, message: Loggable): void;
  info(message: Loggable): void;
  info(...args: Loggable[]) {
    if (args.length > 1) {
      this._log("info", args[1], args[0]);
    } else {
      // only message is passed
      this._log("info", args[0], undefined);
    }
  }

  error(label: Loggable, message: Loggable): void;
  error(message: Loggable): void;
  error(...args: Loggable[]) {
    if (args.length > 1) {
      this._log("error", args[1], args[0]);
    } else {
      // only message is passed
      this._log("error", args[0], undefined);
    }
  }

  wrap<X>(name: string, cb: () => X): X {
    // NOTE: We don't currently implement anything that can utilize `als`
    const ctx = als.getStore();
    const newCtx = [...(ctx ?? []), name];
    return als.run(newCtx, cb) as X;
  }

  static specific(name: string, level?: LogLevel) {
    return new Logger(name, level);
  }

  private _log(level: LogLevel, message: Loggable, label: Loggable | undefined) {
    const minLogLevel = this._forceLevel ?? LOG_CATEGORIES[this._category || ""] ?? "silly";
    if (Logger.compareLevels(level, minLogLevel)) {
      this._do_log(level, message, label);
    }
  }

  protected _do_log(level: LogLevel, message: Loggable, label: Loggable | undefined): void {
    const msgObj =
      message instanceof Error || typeof message === "string"
        ? message
        : util.inspect(message, { depth: 10, colors: true });
    winstonLogger.log(level, msgObj as any, { label });
  }

  private static compareLevels(a: LogLevel, b: LogLevel): boolean {
    const levels: LogLevel[] = ["error", "warn", "info", "verbose", "debug", "silly"];
    return levels.indexOf(a) - levels.indexOf(b) <= 0;
  }
}

const logger = new Logger();
export default logger;

/**
 * Helper functions.
 */

function coloredLevel(level: string): string {
  const l = chalk.bold(chalk.white(` ${level.toUpperCase().padEnd(5)} `));
  switch (level) {
    case "silly": {
      return chalk.bgYellow(l);
    }
    case "debug": {
      return chalk.bgBlue(l);
    }
    case "info": {
      return chalk.bgGreen(l);
    }
    case "error": {
      return chalk.bgRed(l);
    }
    default: {
      return level;
    }
  }
}

function coloredTs(ts: string): string {
  return chalk.bold(`[${ts}]`);
}

function coloredStack(stack: string): string {
  if (!stack) {
    return "";
  }
  const [err, ...s] = stack.split("\n").map((line) => `${_.repeat("\t", 2)}${line}`);
  return `${chalk.bold(chalk.red(err))}\n${chalk.gray(s.join("\n"))}`;
}

function coloredLabel(label: string | undefined): string {
  return label ? chalk.bgGray(chalk.yellowBright(chalk.bold(`[${label}]`))) : "";
}

function padMessage(msg: string): string {
  return msg
    .split("\n")
    .map((line, index) => (index > 0 ? `\t\t${line}` : line))
    .join("\n");
}
