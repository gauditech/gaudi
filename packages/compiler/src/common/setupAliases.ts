import path from "path";

import { addAliases } from "module-alias";

/**
 * NOTE: Make sure to keep aliases in sync with `tsconfig.json` and `jest.config.js`.
 */

addAliases({
  "@compiler": path.join(__dirname, ".."),
});
