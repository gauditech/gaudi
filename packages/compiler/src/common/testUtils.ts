import { compileToAST } from "@compiler/compiler";
import { compilerErrorsToString } from "@compiler/compiler/compilerError";
import { migrate } from "@compiler/compiler/migrate";
import { compose } from "@compiler/composer/composer";
import { Definition } from "@compiler/types/definition";

/**
 * Helper function that compiles directly to definition. This is used in tests.
 */
export function compileFromString(source: string): Definition {
  const inputs = [{ source }];
  const { ast, errors } = compileToAST([{ source }]);
  if (errors) {
    const errorString = compilerErrorsToString(inputs, errors);
    throw new Error(`Failed to compile:\n${errorString}`);
  }
  return compose(migrate(ast));
}
