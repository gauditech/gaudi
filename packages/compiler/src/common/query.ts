import { assertUnreachable, ensureEqual } from "@compiler/common/utils";
import { SelectDef, SelectItem, TypedExprDef } from "@compiler/types/definition";
import _ from "lodash";
import { match } from "ts-pattern";

export function transformSelectPath(select: SelectDef, from: string[], to: string[]): SelectDef {
  return select.map((item: SelectItem): SelectItem => {
    return match(item)
      .with({ kind: "expression" }, (item) => ({
        ...item,
        expr: transformExpressionPaths(item.expr, from, to),
      }))
      .with({ kind: "model-hook" }, (item) => ({
        ...item,
        namePath: transformNamePath(item.namePath, from, to),
      }))
      .with({ kind: "nested-select" }, (item) => {
        return {
          ...item,
          namePath: transformNamePath(item.namePath, from, to),
          select: transformSelectPath(item.select, from, to),
        };
      })
      .exhaustive();
  });
}

export function transformNamePath(path: string[], from: string[], to: string[]): string[] {
  ensureEqual(
    _.isEqual(from, _.take(path, from.length)),
    true,
    `Cannot transform name path: ${path.join(".")} doesn't start with ${from.join(".")}`
  );
  return [...to, ..._.drop(path, from.length)];
}

export function transformNamePaths(paths: string[][], from: string[], to: string[]): string[][] {
  return paths.map((path) => transformNamePath(path, from, to));
}

export function transformExpressionPaths(
  exp: TypedExprDef,
  from: string[],
  to: string[]
): TypedExprDef {
  if (exp === undefined) {
    return undefined;
  }
  switch (exp.kind) {
    case "literal":
    case "variable": {
      return exp;
    }
    case "alias": {
      return { ...exp, namePath: transformNamePath(exp.namePath, from, to) };
    }
    case "function": {
      return {
        ...exp,
        args: exp.args.map((arg) => transformExpressionPaths(arg, from, to)),
      };
    }
    case "aggregate-function": {
      return { ...exp, sourcePath: transformNamePath(exp.sourcePath, from, to) };
    }
    case "in-subquery": {
      return {
        ...exp,
        sourcePath: transformNamePath(exp.sourcePath, from, to),
        lookupExpression: transformExpressionPaths(exp, from, to),
      };
    }
    case "array": {
      return {
        ...exp,
        elements: exp.elements.map((arg) => transformExpressionPaths(arg, from, to)),
      };
    }
    default: {
      assertUnreachable(exp);
    }
  }
}
