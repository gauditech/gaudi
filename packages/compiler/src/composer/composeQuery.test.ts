import { compileFromString } from "@compiler/common/testUtils";
import { CustomOneEndpointDef, UpdateEndpointDef } from "@compiler/types/definition";

describe("compose model queries", () => {
  it("nested example without filters", () => {
    const bp = `
    model Org {
      relation repos { from Repo, through org }
      query back_to_org { from repos.org }
    }
    model Repo {
      reference org { to Org }
    }
    `;
    const def = compileFromString(bp);
    expect(def.models[0].queries).toMatchSnapshot();
  });
  it("example with nested filters", () => {
    const bp = `
    model Org {
      relation repos { from Repo, through org }
      query repos_if_one { from repos, filter { org.id is 1 and is_active } }
    }
    model Repo {
      field is_active { type boolean }
      reference org { to Org }
    }
    `;
    const def = compileFromString(bp);

    expect(def.models[0].queries).toMatchSnapshot();
  });

  it("order and limit", () => {
    const bp = `
    model Org {
      relation repos { from Repo, through org }
      query recent_repos {
        from repos,
        order by { id desc },
        limit 5
      }
    }
    model Repo {
      reference org { to Org }
    }
    `;
    const def = compileFromString(bp);
    expect(def.models[0].queries[0]).toMatchSnapshot();
  });
});

describe("compose action queries", () => {
  describe("actions 2", () => {
    // diffrent "query" action types
    const actions = [
      // --- query action
      // target
      {
        name: "'query' action on target",
        action: `query { from Org, filter { id is 1 }, select {name} }`, // TODO: read from ctx - id
      },
      // other model
      {
        name: "'query' action on other model",
        action: `query { from Repo, filter { id is 1 } }`, // TODO: read from ctx - id
      },

      // --- update action
      // target
      {
        name: "'update' action on target",
        action: `query { from Org as uo, filter { uo.id is 1 }, select {name}, update { set description uo.description + " [updated]" }}`,
      },
      // other model
      {
        name: "'update' action on other model",
        action: `query { from Repo as ur, filter { ur.id is 1 }, update { set name ur.name + " [update]" }}`,
      },

      // --- delete action
      // target
      {
        name: "'delete' action on target",
        action: `query { from Org, filter { description is "test content" }, delete }`,
      },
      // other model
      {
        name: "'delete' action on other model",
        action: `query { from Repo, filter { id is 1 }, delete }`,
      },
    ];

    const bpDefaultEndpoint = (action: string) => `
      model Org {
        field name { type string }
        field description { type string }
      }
      model Repo {
        field name { type string }
      }

      api {
        entrypoint Org {
          // test in native endpoint
          update endpoint {
            action {
              // required default action
              update {
                input { name, description }
              }

              ${action}
            }
          }
        }
      }
    `;

    const bpCustomEndpoint = (action: string) => `
      model Org {
        field name { type string }
        field description { type string }
      }
      model Repo {
        field name { type string }
      }

      api {
        entrypoint Org {

          // test in custom endpoint
          custom endpoint {
            path "customPath"
            method POST
            cardinality one

            action {
              ${action}
            }
          }
        }
      }
    `;

    // iterate actions through different endpoint types
    actions.forEach((a) => {
      describe("default endpoint", () => {
        it(a.name, () => {
          const def = compileFromString(bpDefaultEndpoint(a.action));
          const ep = def.apis[0].entrypoints[0].endpoints[0] as UpdateEndpointDef;
          // take the second action after the default one
          expect(ep.actions[1]).toMatchSnapshot();
        });
      });

      describe("custom endpoint", () => {
        it(a.name, () => {
          const def = compileFromString(bpCustomEndpoint(a.action));
          const ep = def.apis[0].entrypoints[0].endpoints[0] as UpdateEndpointDef;

          // take the first action
          expect(ep.actions[0]).toMatchSnapshot();
        });
      });
    });
  });
});
