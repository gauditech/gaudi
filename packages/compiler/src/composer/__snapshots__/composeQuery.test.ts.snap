// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compose action queries actions 2 custom endpoint 'delete' action on other model 1`] = `
{
  "alias": "$action_0",
  "kind": "query-delete",
  "model": "Repo",
  "query": {
    "filter": {
      "args": [
        {
          "kind": "alias",
          "namePath": [
            "Repo",
            "id",
          ],
        },
        {
          "kind": "literal",
          "literal": {
            "kind": "integer",
            "value": 1,
          },
        },
      ],
      "kind": "function",
      "name": "is",
    },
    "fromAlias": undefined,
    "fromPath": [
      "Repo",
    ],
    "kind": "query",
    "limit": undefined,
    "modelRefKey": "Repo",
    "name": "$query",
    "offset": undefined,
    "orderBy": undefined,
    "refKey": "N/A",
    "retCardinality": "collection",
    "retType": "Repo",
    "select": [
      {
        "alias": "id",
        "expr": {
          "kind": "alias",
          "namePath": [
            "Repo",
            "id",
          ],
        },
        "kind": "expression",
        "type": {
          "kind": "integer",
          "nullable": false,
        },
      },
    ],
  },
}
`;

exports[`compose action queries actions 2 custom endpoint 'delete' action on target 1`] = `
{
  "alias": "$action_0",
  "kind": "query-delete",
  "model": "Org",
  "query": {
    "filter": {
      "args": [
        {
          "kind": "alias",
          "namePath": [
            "Org",
            "description",
          ],
        },
        {
          "kind": "literal",
          "literal": {
            "kind": "string",
            "value": "test content",
          },
        },
      ],
      "kind": "function",
      "name": "is",
    },
    "fromAlias": undefined,
    "fromPath": [
      "Org",
    ],
    "kind": "query",
    "limit": undefined,
    "modelRefKey": "Org",
    "name": "$query",
    "offset": undefined,
    "orderBy": undefined,
    "refKey": "N/A",
    "retCardinality": "collection",
    "retType": "Org",
    "select": [
      {
        "alias": "id",
        "expr": {
          "kind": "alias",
          "namePath": [
            "Org",
            "id",
          ],
        },
        "kind": "expression",
        "type": {
          "kind": "integer",
          "nullable": false,
        },
      },
    ],
  },
}
`;

exports[`compose action queries actions 2 custom endpoint 'query' action on other model 1`] = `
{
  "alias": "$action_0",
  "kind": "query-select",
  "model": "Repo",
  "query": {
    "filter": {
      "args": [
        {
          "kind": "alias",
          "namePath": [
            "Repo",
            "id",
          ],
        },
        {
          "kind": "literal",
          "literal": {
            "kind": "integer",
            "value": 1,
          },
        },
      ],
      "kind": "function",
      "name": "is",
    },
    "fromAlias": undefined,
    "fromPath": [
      "Repo",
    ],
    "kind": "query",
    "limit": undefined,
    "modelRefKey": "Repo",
    "name": "$query",
    "offset": undefined,
    "orderBy": undefined,
    "refKey": "N/A",
    "retCardinality": "collection",
    "retType": "Repo",
    "select": [
      {
        "alias": "id",
        "expr": {
          "kind": "alias",
          "namePath": [
            "Repo",
            "id",
          ],
        },
        "kind": "expression",
        "type": {
          "kind": "integer",
          "nullable": false,
        },
      },
    ],
  },
}
`;

exports[`compose action queries actions 2 custom endpoint 'query' action on target 1`] = `
{
  "alias": "$action_0",
  "kind": "query-select",
  "model": "Org",
  "query": {
    "filter": {
      "args": [
        {
          "kind": "alias",
          "namePath": [
            "Org",
            "id",
          ],
        },
        {
          "kind": "literal",
          "literal": {
            "kind": "integer",
            "value": 1,
          },
        },
      ],
      "kind": "function",
      "name": "is",
    },
    "fromAlias": undefined,
    "fromPath": [
      "Org",
    ],
    "kind": "query",
    "limit": undefined,
    "modelRefKey": "Org",
    "name": "$query",
    "offset": undefined,
    "orderBy": undefined,
    "refKey": "N/A",
    "retCardinality": "collection",
    "retType": "Org",
    "select": [
      {
        "alias": "name",
        "expr": {
          "kind": "alias",
          "namePath": [
            "Org",
            "name",
          ],
        },
        "kind": "expression",
        "type": {
          "kind": "string",
          "nullable": false,
        },
      },
    ],
  },
}
`;

exports[`compose action queries actions 2 custom endpoint 'update' action on other model 1`] = `
{
  "alias": "$action_0",
  "changeset": [
    {
      "name": "name",
      "setter": {
        "args": [
          {
            "kind": "reference-value",
            "target": {
              "access": [
                "name",
              ],
              "alias": "ur",
            },
          },
          {
            "kind": "literal",
            "literal": {
              "kind": "string",
              "value": " [update]",
            },
          },
        ],
        "kind": "function",
        "name": "concat",
      },
    },
  ],
  "kind": "query-update",
  "model": "Repo",
  "query": {
    "filter": {
      "args": [
        {
          "kind": "alias",
          "namePath": [
            "Repo",
            "id",
          ],
        },
        {
          "kind": "literal",
          "literal": {
            "kind": "integer",
            "value": 1,
          },
        },
      ],
      "kind": "function",
      "name": "is",
    },
    "fromAlias": [
      "ur",
    ],
    "fromPath": [
      "Repo",
    ],
    "kind": "query",
    "limit": undefined,
    "modelRefKey": "Repo",
    "name": "$query",
    "offset": undefined,
    "orderBy": undefined,
    "refKey": "N/A",
    "retCardinality": "collection",
    "retType": "Repo",
    "select": [
      {
        "alias": "id",
        "expr": {
          "kind": "alias",
          "namePath": [
            "Repo",
            "id",
          ],
        },
        "kind": "expression",
        "type": {
          "kind": "integer",
          "nullable": false,
        },
      },
    ],
  },
}
`;

exports[`compose action queries actions 2 custom endpoint 'update' action on target 1`] = `
{
  "alias": "$action_0",
  "changeset": [
    {
      "name": "description",
      "setter": {
        "args": [
          {
            "kind": "reference-value",
            "target": {
              "access": [
                "description",
              ],
              "alias": "uo",
            },
          },
          {
            "kind": "literal",
            "literal": {
              "kind": "string",
              "value": " [updated]",
            },
          },
        ],
        "kind": "function",
        "name": "concat",
      },
    },
  ],
  "kind": "query-update",
  "model": "Org",
  "query": {
    "filter": {
      "args": [
        {
          "kind": "alias",
          "namePath": [
            "Org",
            "id",
          ],
        },
        {
          "kind": "literal",
          "literal": {
            "kind": "integer",
            "value": 1,
          },
        },
      ],
      "kind": "function",
      "name": "is",
    },
    "fromAlias": [
      "uo",
    ],
    "fromPath": [
      "Org",
    ],
    "kind": "query",
    "limit": undefined,
    "modelRefKey": "Org",
    "name": "$query",
    "offset": undefined,
    "orderBy": undefined,
    "refKey": "N/A",
    "retCardinality": "collection",
    "retType": "Org",
    "select": [
      {
        "alias": "name",
        "expr": {
          "kind": "alias",
          "namePath": [
            "Org",
            "name",
          ],
        },
        "kind": "expression",
        "type": {
          "kind": "string",
          "nullable": false,
        },
      },
    ],
  },
}
`;

exports[`compose action queries actions 2 default endpoint 'delete' action on other model 1`] = `
{
  "alias": "$action_1",
  "kind": "query-delete",
  "model": "Repo",
  "query": {
    "filter": {
      "args": [
        {
          "kind": "alias",
          "namePath": [
            "Repo",
            "id",
          ],
        },
        {
          "kind": "literal",
          "literal": {
            "kind": "integer",
            "value": 1,
          },
        },
      ],
      "kind": "function",
      "name": "is",
    },
    "fromAlias": undefined,
    "fromPath": [
      "Repo",
    ],
    "kind": "query",
    "limit": undefined,
    "modelRefKey": "Repo",
    "name": "$query",
    "offset": undefined,
    "orderBy": undefined,
    "refKey": "N/A",
    "retCardinality": "collection",
    "retType": "Repo",
    "select": [
      {
        "alias": "id",
        "expr": {
          "kind": "alias",
          "namePath": [
            "Repo",
            "id",
          ],
        },
        "kind": "expression",
        "type": {
          "kind": "integer",
          "nullable": false,
        },
      },
    ],
  },
}
`;

exports[`compose action queries actions 2 default endpoint 'delete' action on target 1`] = `
{
  "alias": "$action_1",
  "kind": "query-delete",
  "model": "Org",
  "query": {
    "filter": {
      "args": [
        {
          "kind": "alias",
          "namePath": [
            "Org",
            "description",
          ],
        },
        {
          "kind": "literal",
          "literal": {
            "kind": "string",
            "value": "test content",
          },
        },
      ],
      "kind": "function",
      "name": "is",
    },
    "fromAlias": undefined,
    "fromPath": [
      "Org",
    ],
    "kind": "query",
    "limit": undefined,
    "modelRefKey": "Org",
    "name": "$query",
    "offset": undefined,
    "orderBy": undefined,
    "refKey": "N/A",
    "retCardinality": "collection",
    "retType": "Org",
    "select": [
      {
        "alias": "id",
        "expr": {
          "kind": "alias",
          "namePath": [
            "Org",
            "id",
          ],
        },
        "kind": "expression",
        "type": {
          "kind": "integer",
          "nullable": false,
        },
      },
    ],
  },
}
`;

exports[`compose action queries actions 2 default endpoint 'query' action on other model 1`] = `
{
  "alias": "$action_1",
  "kind": "query-select",
  "model": "Repo",
  "query": {
    "filter": {
      "args": [
        {
          "kind": "alias",
          "namePath": [
            "Repo",
            "id",
          ],
        },
        {
          "kind": "literal",
          "literal": {
            "kind": "integer",
            "value": 1,
          },
        },
      ],
      "kind": "function",
      "name": "is",
    },
    "fromAlias": undefined,
    "fromPath": [
      "Repo",
    ],
    "kind": "query",
    "limit": undefined,
    "modelRefKey": "Repo",
    "name": "$query",
    "offset": undefined,
    "orderBy": undefined,
    "refKey": "N/A",
    "retCardinality": "collection",
    "retType": "Repo",
    "select": [
      {
        "alias": "id",
        "expr": {
          "kind": "alias",
          "namePath": [
            "Repo",
            "id",
          ],
        },
        "kind": "expression",
        "type": {
          "kind": "integer",
          "nullable": false,
        },
      },
    ],
  },
}
`;

exports[`compose action queries actions 2 default endpoint 'query' action on target 1`] = `
{
  "alias": "$action_1",
  "kind": "query-select",
  "model": "Org",
  "query": {
    "filter": {
      "args": [
        {
          "kind": "alias",
          "namePath": [
            "Org",
            "id",
          ],
        },
        {
          "kind": "literal",
          "literal": {
            "kind": "integer",
            "value": 1,
          },
        },
      ],
      "kind": "function",
      "name": "is",
    },
    "fromAlias": undefined,
    "fromPath": [
      "Org",
    ],
    "kind": "query",
    "limit": undefined,
    "modelRefKey": "Org",
    "name": "$query",
    "offset": undefined,
    "orderBy": undefined,
    "refKey": "N/A",
    "retCardinality": "collection",
    "retType": "Org",
    "select": [
      {
        "alias": "name",
        "expr": {
          "kind": "alias",
          "namePath": [
            "Org",
            "name",
          ],
        },
        "kind": "expression",
        "type": {
          "kind": "string",
          "nullable": false,
        },
      },
    ],
  },
}
`;

exports[`compose action queries actions 2 default endpoint 'update' action on other model 1`] = `
{
  "alias": "$action_1",
  "changeset": [
    {
      "name": "name",
      "setter": {
        "args": [
          {
            "kind": "reference-value",
            "target": {
              "access": [
                "name",
              ],
              "alias": "ur",
            },
          },
          {
            "kind": "literal",
            "literal": {
              "kind": "string",
              "value": " [update]",
            },
          },
        ],
        "kind": "function",
        "name": "concat",
      },
    },
  ],
  "kind": "query-update",
  "model": "Repo",
  "query": {
    "filter": {
      "args": [
        {
          "kind": "alias",
          "namePath": [
            "Repo",
            "id",
          ],
        },
        {
          "kind": "literal",
          "literal": {
            "kind": "integer",
            "value": 1,
          },
        },
      ],
      "kind": "function",
      "name": "is",
    },
    "fromAlias": [
      "ur",
    ],
    "fromPath": [
      "Repo",
    ],
    "kind": "query",
    "limit": undefined,
    "modelRefKey": "Repo",
    "name": "$query",
    "offset": undefined,
    "orderBy": undefined,
    "refKey": "N/A",
    "retCardinality": "collection",
    "retType": "Repo",
    "select": [
      {
        "alias": "id",
        "expr": {
          "kind": "alias",
          "namePath": [
            "Repo",
            "id",
          ],
        },
        "kind": "expression",
        "type": {
          "kind": "integer",
          "nullable": false,
        },
      },
    ],
  },
}
`;

exports[`compose action queries actions 2 default endpoint 'update' action on target 1`] = `
{
  "alias": "$action_1",
  "changeset": [
    {
      "name": "description",
      "setter": {
        "args": [
          {
            "kind": "reference-value",
            "target": {
              "access": [
                "description",
              ],
              "alias": "uo",
            },
          },
          {
            "kind": "literal",
            "literal": {
              "kind": "string",
              "value": " [updated]",
            },
          },
        ],
        "kind": "function",
        "name": "concat",
      },
    },
  ],
  "kind": "query-update",
  "model": "Org",
  "query": {
    "filter": {
      "args": [
        {
          "kind": "alias",
          "namePath": [
            "Org",
            "id",
          ],
        },
        {
          "kind": "literal",
          "literal": {
            "kind": "integer",
            "value": 1,
          },
        },
      ],
      "kind": "function",
      "name": "is",
    },
    "fromAlias": [
      "uo",
    ],
    "fromPath": [
      "Org",
    ],
    "kind": "query",
    "limit": undefined,
    "modelRefKey": "Org",
    "name": "$query",
    "offset": undefined,
    "orderBy": undefined,
    "refKey": "N/A",
    "retCardinality": "collection",
    "retType": "Org",
    "select": [
      {
        "alias": "name",
        "expr": {
          "kind": "alias",
          "namePath": [
            "Org",
            "name",
          ],
        },
        "kind": "expression",
        "type": {
          "kind": "string",
          "nullable": false,
        },
      },
    ],
  },
}
`;

exports[`compose model queries example with nested filters 1`] = `
[
  {
    "filter": {
      "args": [
        {
          "args": [
            {
              "kind": "alias",
              "namePath": [
                "Org",
                "repos",
                "org",
                "id",
              ],
            },
            {
              "kind": "literal",
              "literal": {
                "kind": "integer",
                "value": 1,
              },
            },
          ],
          "kind": "function",
          "name": "is",
        },
        {
          "kind": "alias",
          "namePath": [
            "Org",
            "repos",
            "is_active",
          ],
        },
      ],
      "kind": "function",
      "name": "and",
    },
    "fromAlias": undefined,
    "fromPath": [
      "Org",
      "repos",
    ],
    "kind": "query",
    "limit": undefined,
    "modelRefKey": "Org",
    "name": "repos_if_one",
    "offset": undefined,
    "orderBy": undefined,
    "refKey": "Org.repos_if_one",
    "retCardinality": "collection",
    "retType": "Repo",
    "select": [],
  },
]
`;

exports[`compose model queries nested example without filters 1`] = `
[
  {
    "filter": undefined,
    "fromAlias": undefined,
    "fromPath": [
      "Org",
      "repos",
      "org",
    ],
    "kind": "query",
    "limit": undefined,
    "modelRefKey": "Org",
    "name": "back_to_org",
    "offset": undefined,
    "orderBy": undefined,
    "refKey": "Org.back_to_org",
    "retCardinality": "collection",
    "retType": "Org",
    "select": [],
  },
]
`;

exports[`compose model queries order and limit 1`] = `
{
  "filter": undefined,
  "fromAlias": undefined,
  "fromPath": [
    "Org",
    "repos",
  ],
  "kind": "query",
  "limit": 5,
  "modelRefKey": "Org",
  "name": "recent_repos",
  "offset": undefined,
  "orderBy": [
    {
      "direction": "desc",
      "exp": {
        "kind": "alias",
        "namePath": [
          "Org",
          "repos",
          "id",
        ],
      },
    },
  ],
  "refKey": "Org.recent_repos",
  "retCardinality": "collection",
  "retType": "Repo",
  "select": [],
}
`;
